spark/src/main/java/org/apache/iceberg/spark/data/vectorized/ArrowVectorAccessors.java

```diff
@@ -0,0 +1,495 @@
+        return new TimestampAccessor((TimeStampMicroTZVector) vector);
+      } else if (vector instanceof ListVector) {
+        ListVector listVector = (ListVector) vector;
+        return new ArrayAccessor(listVector);
+      } else if (vector instanceof StructVector) {
+        StructVector structVector = (StructVector) vector;
+        return new StructAccessor(structVector);
+      }
+    }
+    throw new UnsupportedOperationException("Unsupported type: " + primitive);
+  }
+
+  private static class BooleanAccessor extends ArrowVectorAccessor {
+
+    private final BitVector vector;
+
+    BooleanAccessor(BitVector vector) {
+      super(vector);
+      this.vector = vector;
+    }
+
+    @Override
+    final boolean getBoolean(int rowId) {
+      return vector.get(rowId) == 1;
+    }
+  }
+
+  private static class IntAccessor extends ArrowVectorAccessor {
+
+    private final IntVector vector;
+
+    IntAccessor(IntVector vector) {
+      super(vector);
+      this.vector = vector;
+    }
+
+    @Override
+    final int getInt(int rowId) {
+      return vector.get(rowId);
+    }
+  }
+
+  private static class LongAccessor extends ArrowVectorAccessor {
+
+    private final BigIntVector vector;
+
+    LongAccessor(BigIntVector vector) {
+      super(vector);
+      this.vector = vector;
+    }
+
+    @Override
+    final long getLong(int rowId) {
+      return vector.get(rowId);
+    }
+  }
+
+  private static class DictionaryLongAccessor extends DictionaryArrowVectorAccessor {
+
+    private final IntVector vector;
+
+    DictionaryLongAccessor(IntVector vector, Dictionary dictionary) {
+      super(vector, dictionary);
+      this.vector = vector;
+    }
+
+    @Override
+    final long getLong(int rowId) {
+      return parquetDictionary.decodeToLong(vector.get(rowId));
+    }
+  }
+
+  private static class FloatAccessor extends ArrowVectorAccessor {
+
+    private final Float4Vector vector;
+
+    FloatAccessor(Float4Vector vector) {
+      super(vector);
+      this.vector = vector;
+    }
+
+    @Override
+    final float getFloat(int rowId) {
+      return vector.get(rowId);
+    }
+  }
+
+  private static class DictionaryFloatAccessor extends DictionaryArrowVectorAccessor {
+
+    private final IntVector vector;
+
+    DictionaryFloatAccessor(IntVector vector, Dictionary dictionary) {
+      super(vector, dictionary);
+      this.vector = vector;
+    }
+
+    @Override
+    final float getFloat(int rowId) {
+      return parquetDictionary.decodeToFloat(vector.get(rowId));
+    }
+  }
+
+  private static class DoubleAccessor extends ArrowVectorAccessor {
+
+    private final Float8Vector vector;
+
+    DoubleAccessor(Float8Vector vector) {
+      super(vector);
+      this.vector = vector;
+    }
+
+    @Override
+    final double getDouble(int rowId) {
+      return vector.get(rowId);
+    }
+  }
+
+  private static class DictionaryDoubleAccessor extends DictionaryArrowVectorAccessor {
```

Comment:
Similar to the string dictionary accessor, I think we can avoid extra work and make this simpler by eagerly decoding here:

```java
  private static class DictionaryDoubleAccessor extends ArrowVectorAccessor {
    private final IntVector vector;
    private final double[] decodedDictionary;

    DictionaryDoubleAccessor(IntVector vector, Dictionary dictionary) {
      super(vector);
      this.vector = vector;
      this.decodedDictionary = IntStream.rangeClosed(0, dictionary.getMaxId())
          .mapToDouble(dictionary::decodeToDouble)
          .toArray();
    }

    @Override
    final double getDouble(int rowId) {
      return decodedDictionary[vector.get(rowId)];
    }
  }
```

Parquet decodes into a `double[]` as well, but this avoids dynamic dispatch costs for `decodeToDouble` and is easier to see what's happening.